iOS/MMIWG2S/ViewControllers/FacesARViewController.swift

```diff
@@ -0,0 +1,159 @@
+        sceneView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
+        sceneView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
+    }
+    
+    /// Setup the ui used during the face session
+    private func setupFaceViewer() {
+        faceViewer.delegate = self
+        
+        view.addSubview(faceViewer)
+        faceViewer.translatesAutoresizingMaskIntoConstraints = false
+        faceViewer.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
+        faceViewer.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
+        faceViewer.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
+        faceViewer.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
+    }
+    
+    private func resetTracking() {
+        guard ARFaceTrackingConfiguration.isSupported else { return }
+        sceneView.session.run(configuration, options: [.resetTracking, .removeExistingAnchors])
+    }
+    
+}
+
+// MARK: - ARFaceViewerUIDelegate
+
+extension FacesARViewController: ARFaceViewerUIDelegate {
+    
+    func textureChanged(toRed: Bool) {
+        guard let contentNode = sceneView.scene.rootNode.childNode(withName: faceNodeName, recursively: true) else { return }
+        if toRed {
+            contentNode.geometry?.materials.first?.diffuse.contents = imageRed
+        } else {
+            contentNode.geometry?.materials.first?.diffuse.contents = imageBlack
+        }
+    }
+    
+    func generateShareImage() -> UIImage? {
+        let image = sceneView.snapshot()
+        sceneView.session.pause()
+        return image
+    }
+    
+    func resumeSessionOnceCaptured() {
+        resetTracking()
+    }
+    
+}
+
+// MARK: - ARSessionDelegate
+
+extension FacesARViewController: ARSessionDelegate {
+    
+    func session(_ session: ARSession, didFailWithError error: Error) {
+        guard error is ARError else { return }
+        
+        let errorWithInfo = error as NSError
+        let messages = [
+            errorWithInfo.localizedDescription,
+            errorWithInfo.localizedFailureReason,
+            errorWithInfo.localizedRecoverySuggestion
+        ]
+        let errorMessage = messages.compactMap({ $0 }).joined(separator: "\n")
+        
+        DispatchQueue.main.async { [weak self] in
+            self?.displayError(.errorARSessionTitle, message: errorMessage, completion: self?.resetTracking)
+        }
+    }
+    
+}
+
+extension FacesARViewController: ARSCNViewDelegate {
+    
+    func renderer(_ renderer: SCNSceneRenderer, didAdd node: SCNNode, for anchor: ARAnchor) {
+        guard let faceAnchor = anchor as? ARFaceAnchor else { return }
+        currentFaceAnchor = faceAnchor
+        
+        if node.childNodes.isEmpty,
+           let device = sceneView.device,
+           let faceGeometry = ARSCNFaceGeometry(device: device),
+           let material = faceGeometry.firstMaterial {
```

Comment:
This is really unique. Is the if statement checking on the first and then setting lets to be used? Or is each let part of the if check?

Reply:
They're all treated as `if let`